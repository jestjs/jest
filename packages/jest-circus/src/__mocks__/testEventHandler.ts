/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {Circus} from '@jest/types';

const testEventHandler: Circus.EventHandler = (event, state) => {
  switch (event.name) {
    case 'start_describe_definition':
    case 'finish_describe_definition': {
      console.log(`${event.name}:`, event.blockName);
      break;
    }
    case 'run_describe_start':
    case 'run_describe_finish': {
      console.log(`${event.name}:`, event.describeBlock.name);
      break;
    }
    case 'test_start':
    case 'test_started':
    case 'test_retry':
    case 'test_done': {
      console.log(`${event.name}:`, event.test.name);
      break;
    }

    case 'add_test': {
      console.log(`${event.name}:`, event.testName);
      break;
    }

    case 'test_fn_start':
    case 'test_fn_success':
    case 'test_fn_failure': {
      console.log(`${event.name}:`, event.test.name);
      break;
    }

    case 'add_hook': {
      console.log(`${event.name}:`, event.hookType);
      break;
    }

    case 'hook_start':
    case 'hook_success':
    case 'hook_failure': {
      console.log(`${event.name}:`, event.hook.type);
      break;
    }

    default:
      console.log(event.name);
  }

  if (event.name === 'run_finish') {
    console.log('');
    console.log(`unhandledErrors: ${String(state.unhandledErrors.length)}`);
  }
};

export default testEventHandler;
