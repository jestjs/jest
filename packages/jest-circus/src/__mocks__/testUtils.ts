/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {sync as spawnSync} from 'execa';
import * as fs from 'graceful-fs';
import tempy = require('tempy');

const CIRCUS_PATH = require.resolve('../').replaceAll('\\', '\\\\');
const CIRCUS_RUN_PATH = require.resolve('../run').replaceAll('\\', '\\\\');
const CIRCUS_STATE_PATH = require.resolve('../state').replaceAll('\\', '\\\\');
const TEST_EVENT_HANDLER_PATH = require
  .resolve('./testEventHandler')
  .replaceAll('\\', '\\\\');
const BABEL_REGISTER_PATH = require
  .resolve('@babel/register')
  .replaceAll('\\', '\\\\');

export const runTest = (
  source: string,
  opts?: {seed?: number; randomize?: boolean},
) => {
  const tmpFilename = tempy.file();

  const content = `
    require('${BABEL_REGISTER_PATH}')({extensions: [".js", ".ts"]});
    const circus = require('${CIRCUS_PATH}');
    global.test = circus.test;
    global.describe = circus.describe;
    global.beforeEach = circus.beforeEach;
    global.afterEach = circus.afterEach;
    global.beforeAll = circus.beforeAll;
    global.afterAll = circus.afterAll;

    const testEventHandler = require('${TEST_EVENT_HANDLER_PATH}').default;
    const {addEventHandler, removeEventHandler, getState} = require('${CIRCUS_STATE_PATH}');
    getState().randomize = ${opts?.randomize};
    getState().seed = ${opts?.seed ?? 0};
    addEventHandler(testEventHandler);

    ${source};

    const run = require('${CIRCUS_RUN_PATH}').default;

    run();
  `;

  fs.writeFileSync(tmpFilename, content);
  const result = spawnSync('node', [tmpFilename], {
    cwd: process.cwd(),
  });

  if (result.exitCode !== 0) {
    const message = `
      STDOUT: ${result.stdout && result.stdout.toString()}
      STDERR: ${result.stderr && result.stderr.toString()}
      STATUS: ${result.exitCode}
    `;
    throw new Error(message);
  }

  fs.rmSync(tmpFilename, {force: true});

  if (result.stderr) {
    throw new Error(
      `
      Unexpected stderr:
      ${result.stderr}
    `,
    );
  }
  return result;
};
