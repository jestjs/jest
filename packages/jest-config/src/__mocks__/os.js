/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

const os = jest.createMockFromModule('os');

let cpus;
function __setCpus(newCpus) {
  cpus = newCpus;
}

os.__setCpus = __setCpus;
os.cpus = jest.fn(() => cpus);
os.availableParallelism = jest.fn(() => cpus?.length ?? 0);

module.exports = os;
