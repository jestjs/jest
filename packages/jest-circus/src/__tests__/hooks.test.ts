/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {runTest} from '../__mocks__/testUtils';

test('beforeEach is executed before each test in current/child describe blocks', () => {
  const {stdout} = runTest(`
    describe('describe', () => {
      beforeEach(() => console.log('> describe beforeEach'));
      test('one', () => {});
      test('two', () => {});
      describe('2nd level describe', () => {
        beforeEach(() => console.log('> 2nd level describe beforeEach'));
        test('2nd level test', () => {});

        describe('3rd level describe', () => {
          test('3rd level test', () => {});
          test('3rd level test#2', () => {});
        });
      });
    })

    describe('2nd describe', () => {
      beforeEach(() => {
        console.log('> 2nd describe beforeEach that throws')
        throw new Error('alabama');
      });
      test('2nd describe test', () => {});
    })
  `);

  expect(stdout).toMatchSnapshot();
});

test('multiple before each hooks in one describe are executed in the right order', () => {
  const {stdout} = runTest(`
    describe('describe 1', () => {
      beforeEach(() => {
        console.log('before each 1');
      });
      beforeEach(() => {
        console.log('before each 2');
      });

      describe('2nd level describe', () => {
        test('test', () => {});
      });
    });
  `);

  expect(stdout).toMatchSnapshot();
});

test('beforeAll is executed correctly', () => {
  const {stdout} = runTest(`
    describe('describe 1', () => {
      beforeAll(() => console.log('> beforeAll 1'));
      test('test 1', () => console.log('> test 1'));

      describe('2nd level describe', () => {
        beforeAll(() => console.log('> beforeAll 2'));
        test('test 2', () => console.log('> test 2'));
        test('test 3', () => console.log('> test 3'));
      });
    });
  `);

  expect(stdout).toMatchSnapshot();
});
