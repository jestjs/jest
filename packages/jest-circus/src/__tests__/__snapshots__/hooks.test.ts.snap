// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`beforeAll is exectued correctly 1`] = `
"start_describe_definition: describe 1
add_hook: beforeAll
add_test: test 1
start_describe_definition: 2nd level describe
add_hook: beforeAll
add_test: test 2
add_test: test 3
finish_describe_definition: 2nd level describe
finish_describe_definition: describe 1
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describe 1
hook_start: beforeAll
> beforeAll 1
hook_success: beforeAll
test_start: test 1
test_started: test 1
test_fn_start: test 1
> test 1
test_fn_success: test 1
test_done: test 1
run_describe_start: 2nd level describe
hook_start: beforeAll
> beforeAll 2
hook_success: beforeAll
test_start: test 2
test_started: test 2
test_fn_start: test 2
> test 2
test_fn_success: test 2
test_done: test 2
test_start: test 3
test_started: test 3
test_fn_start: test 3
> test 3
test_fn_success: test 3
test_done: test 3
run_describe_finish: 2nd level describe
run_describe_finish: describe 1
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;

exports[`beforeEach is executed before each test in current/child describe blocks 1`] = `
"start_describe_definition: describe
add_hook: beforeEach
add_test: one
add_test: two
start_describe_definition: 2nd level describe
add_hook: beforeEach
add_test: 2nd level test
start_describe_definition: 3rd level describe
add_test: 3rd level test
add_test: 3rd level test#2
finish_describe_definition: 3rd level describe
finish_describe_definition: 2nd level describe
finish_describe_definition: describe
start_describe_definition: 2nd describe
add_hook: beforeEach
add_test: 2nd describe test
finish_describe_definition: 2nd describe
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describe
test_start: one
test_started: one
hook_start: beforeEach
> describe beforeEach
hook_success: beforeEach
test_fn_start: one
test_fn_success: one
test_done: one
test_start: two
test_started: two
hook_start: beforeEach
> describe beforeEach
hook_success: beforeEach
test_fn_start: two
test_fn_success: two
test_done: two
run_describe_start: 2nd level describe
test_start: 2nd level test
test_started: 2nd level test
hook_start: beforeEach
> describe beforeEach
hook_success: beforeEach
hook_start: beforeEach
> 2nd level describe beforeEach
hook_success: beforeEach
test_fn_start: 2nd level test
test_fn_success: 2nd level test
test_done: 2nd level test
run_describe_start: 3rd level describe
test_start: 3rd level test
test_started: 3rd level test
hook_start: beforeEach
> describe beforeEach
hook_success: beforeEach
hook_start: beforeEach
> 2nd level describe beforeEach
hook_success: beforeEach
test_fn_start: 3rd level test
test_fn_success: 3rd level test
test_done: 3rd level test
test_start: 3rd level test#2
test_started: 3rd level test#2
hook_start: beforeEach
> describe beforeEach
hook_success: beforeEach
hook_start: beforeEach
> 2nd level describe beforeEach
hook_success: beforeEach
test_fn_start: 3rd level test#2
test_fn_success: 3rd level test#2
test_done: 3rd level test#2
run_describe_finish: 3rd level describe
run_describe_finish: 2nd level describe
run_describe_finish: describe
run_describe_start: 2nd describe
test_start: 2nd describe test
test_started: 2nd describe test
hook_start: beforeEach
> 2nd describe beforeEach that throws
hook_failure: beforeEach
test_fn_start: 2nd describe test
test_done: 2nd describe test
run_describe_finish: 2nd describe
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;

exports[`multiple before each hooks in one describe are executed in the right order 1`] = `
"start_describe_definition: describe 1
add_hook: beforeEach
add_hook: beforeEach
start_describe_definition: 2nd level describe
add_test: test
finish_describe_definition: 2nd level describe
finish_describe_definition: describe 1
run_start
run_describe_start: ROOT_DESCRIBE_BLOCK
run_describe_start: describe 1
run_describe_start: 2nd level describe
test_start: test
test_started: test
hook_start: beforeEach
before each 1
hook_success: beforeEach
hook_start: beforeEach
before each 2
hook_success: beforeEach
test_fn_start: test
test_fn_success: test
test_done: test
run_describe_finish: 2nd level describe
run_describe_finish: describe 1
run_describe_finish: ROOT_DESCRIBE_BLOCK
run_finish

unhandledErrors: 0"
`;
